java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/BinaryRelation.java

```diff
@@ -21,336 +21,312 @@
-        relation.inverse.symmetric = new NotEqualRelation(rightOp, leftOp);
-        relation.inverse.inverse = relation;
-        relation.symmetric = new EqualRelation(rightOp, leftOp);
-        relation.symmetric.symmetric = relation;
-        relation.symmetric.inverse = relation.inverse.symmetric;
-        break;
       case NOT_EQUAL:
-        relation = binaryRelation(Kind.EQUAL, leftOp, rightOp).inverse;
-        break;
       case LESS_THAN:
-        relation = new LessThanRelation(leftOp, rightOp);
-        break;
-      case LESS_THAN_OR_EQUAL:
-        relation = new LessThanOrEqualRelation(leftOp, rightOp);
-        break;
-      case GREATER_THAN:
-        relation = new GreaterThanRelation(leftOp, rightOp);
-        break;
       case GREATER_THAN_OR_EQUAL:
-        relation = new GreaterThanOrEqualRelation(leftOp, rightOp);
-        break;
       case METHOD_EQUALS:
-        relation = new MethodEqualsRelation(leftOp, rightOp);
-        break;
       case NOT_METHOD_EQUALS:
-        relation = new NotMethodEqualsRelation(leftOp, rightOp);
-        break;
+        return new BinaryRelation(kind, leftOp, rightOp);
+      case LESS_THAN_OR_EQUAL:
+      case GREATER_THAN:
+        return new BinaryRelation(kind.symmetric(), rightOp, leftOp);
       default:
         throw new IllegalStateException("Creation of relation of kind " + kind + " is missing!");
     }
-    return relation;
-
+  }
 
+  public RelationState resolveState(Collection<BinaryRelation> knownRelations) {
+    if(hasSameOperand()) {
+      return relationStateForSameOperand();
+    }
+    HashSet<BinaryRelation> usedRelations = new HashSet<>();
+    Set<BinaryRelation> relations = new HashSet<>(knownRelations);
+    while (!relations.isEmpty()) {
+      if (usedRelations.size() > 200) {
+        throw new TransitiveRelationExceededException();
+      }
+      for (BinaryRelation relation : relations) {
+        RelationState result = relation.implies(this);
+        if (result.isDetermined()) {
+          return result;
+        }
+        usedRelations.add(relation);
+      }
+      relations = deduce(usedRelations);
+    }
+    return RelationState.UNDETERMINED;
   }
 
+  static Set<BinaryRelation> deduce(Set<BinaryRelation> relations) {
+    Set<BinaryRelation> result = new HashSet<>();
+    BinaryRelation[] binaryRelations = relations.toArray(new BinaryRelation[relations.size()]);
+    for (int i = 0; i < binaryRelations.length; i++) {
+      for (int j = i + 1; j < binaryRelations.length; j++) {
+        BinaryRelation a = binaryRelations[i];
+        BinaryRelation b = binaryRelations[j];
+        BinaryRelation simplified = simplify(a, b);
+        if (simplified != null && !relations.contains(simplified)) {
+          result.add(simplified);
+        }
+        if (a.inTransitiveRelationship(b)) {
+          BinaryRelation transitive = combineTransitively(a, b);
+          if (transitive != null && !relations.contains(transitive)) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png 'Severity: CRITICAL') Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS134)